### PR TITLE
rpctest: refactor package

### PR DESCRIPTION
**Pull request by JFixby**
The goal of this PR is to push towards the resolution of the issues: #1259 and #1262 

Background: 

 - `rpctest`-package was not functional.

 - Test cases were located in one file `rpctest/rpcharness_test.go` and were executed from one entry point, making it a one single test with about of 2300 lines of code.

 - Running multiple harnesses was causing network port collisions

Proposed changes:

 - The test was separated in 16 different tests, each placed into a dedicated file for better code readability and maintenance.
 
- `HarnessPool` was added to manage spawning and disposal of concurrent harnesses

- Overall harness setup code was simplified and delegated to it's components.

 Package structure:

  - `TestMain()` is executed by go-test, and is responsible for setting up
 and disposing test harnesses. See `rpctest_test.go` for details.

  - Test cases are located in `enumerated _test.go` files. Each test runs
 independently. The enumeration specifies test execution order.

  - Test cases retrieve harnesses via `ObtainHarness()` function, which
 redirects it's calls to dedicated harnesses `Pool`

  - `HarnessesPool` as defined in `harnessespool.go` keeps track of reusable
 harness instances.

  - The `Pool` is configured to produce harnesses with a test chain
 of desired height, providing `numMatureOutputs` coinbases to allow
 spending from for testing purposes. See `chainwithmo.go` for details.

I did not test any of that on Linux/MacOS  yet.

### CODE REVIEW

**Comment by JFixby** [comment]
Tests output:
```
=== RUN   TestGetNewAddress
--- FAIL: TestGetNewAddress (0.35s)
	01_getnewaddress_test.go:69: -4: wallet.NewExternalAddress: policy violation: generating
 next address violates the unused address gap limit policy
=== RUN   TestValidateAddress
--- FAIL: TestValidateAddress (0.00s)
	02_validateaddress_test.go:42: -4: wallet.NewExternalAddress: policy violation: generating
 next address violates the unused address gap limit policy
=== RUN   TestWalletPassphrase
--- PASS: TestWalletPassphrase (11.99s)
=== RUN   TestGetBalance
--- PASS: TestGetBalance (1.49s)
=== RUN   TestListAccounts
--- PASS: TestListAccounts (2.91s)
=== RUN   TestListUnspent
--- PASS: TestListUnspent (5.21s)
=== RUN   TestSendToAddress
--- FAIL: TestSendToAddress (0.00s)
	07_sendtoaddress_test.go:26: -4: wallet.NewExternalAddress: policy violation: generating
 next address violates the unused address gap limit policy
=== RUN   TestSendFrom
--- PASS: TestSendFrom (2.50s)
=== RUN   TestSendMany
--- PASS: TestSendMany (1.79s)
	09_sendmany_test.go:95: Raw TX before mining block:  
&{[71 64 94 95 167 156 244 104 32 211 120 159 151 134 29 240 158 35 111 169 218 55 19 21 207 111 92 65 108 144 104 116] 0xc0426a8280 -1 -1}  Fee:  0.000289 DCR
	09_sendmany_test.go:106: Raw TX after mining block:  
&{[71 64 94 95 167 156 244 104 32 211 120 159 151 134 29 240 158 35 111 169 218 55 19 21 207 111 92 65 108 144 104 116] 0xc0421d2230 -1 -1}  Fee:  0.000289 DCR
=== RUN   TestListTransactions
--- FAIL: TestListTransactions (0.09s)
	10_listtransactions_test.go:147: Expected 43 listtransactions results, got 42
=== RUN   TestGetSetRelayFee
--- PASS: TestGetSetRelayFee (1.48s)
	11_getsetrelayfee_test.go:95: Set relay fee: 0.0015, actual: 0.0015059523809523808
=== RUN   TestGetSetTicketFee
--- PASS: TestGetSetTicketFee (2.20s)
	12_getsetticketfee_test.go:97: Set ticket fee: 0.0015, actual: 0.0015202702702702701
=== RUN   TestGetTickets
--- PASS: TestGetTickets (2.53s)
=== RUN   TestPurchaseTickets
--- FAIL: TestPurchaseTickets (0.00s)
	14_purchasetickets_test.go:31: -4: wallet.NewExternalAddress: policy violation: generating
 next address violates the unused address gap limit policy
=== RUN   TestGetStakeInfo
--- FAIL: TestGetStakeInfo (0.00s)
	15_getstakeinfo_test.go:77: Stake difficulty mismatch: 0.000200 vs 0.000000
 (getstakedifficulty, getstakeinfo)
=== RUN   TestWalletInfo
--- PASS: TestWalletInfo (0.00s)
FAIL
```

**Comment by jrick** [comment]
Please `t.Skip("needs update")` the tests that fail, so we can test the ones that do work.

**Comment by JFixby** [comment]
image

**Comment by JFixby** [comment]
I disabled short mode for `TestWalletPassphrase` to see if Travis can build it.
Says ```mkdir /tmp/rpctest935290963/harness-main/dcrd: permission denied```

File: rpctest/10_listtransactions_test.go [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+	// Number of results should be +3 now
+	txListAll, err := wcl.ListTransactionsCount("*", 9999999)
+	if err != nil {
+		t.Fatal("ListTransactionsCount failed:", err)
+	}
+	// Expect 3 more results in the list: a receive for the owned address in
+	// the amount sent, a send in the amount sent, and the a send from the
+	// original outpoint for the mined coins.
+	expectedAdditional := 3
```
Comment (JFixby) [inline rpctest/10_listtransactions_test.go]: Actual:
 ```
(dcrjson.ListTransactionsResult) {
  Account: (string) (len=20) "listTransactionsTest",
  Address: (string) (len=35) "SsdLk37kzNfTtvk4nmxFZawcYvuehiCxAPe",
  Amount: (float64) 2.4,
  BlockHash: (string) "",
  BlockIndex: (*int64)(<nil>),
  BlockTime: (int64) 0,
  Category: (string) (len=7) "receive",
  Confirmations: (int64) 0,
  Fee: (*float64)(<nil>),
  Generated: (bool) false,
  InvolvesWatchOnly: (bool) false,
  Time: (int64) 1535622117,
  TimeReceived: (int64) 1535622117,
  TxID: (string) (len=64) "93676ace2552d5947abdf7ba8e6ca43a4719349dbd23dafcc24b966d5fd6e31d",
  TxType: (*dcrjson.ListTransactionsTxType)(0xc04225c360)((len=7) "regular"),
  Vout: (uint32) 1,
  WalletConflicts: ([]string) {
  },
  Comment: (string) "",
  OtherAccount: (string) ""
 },

(dcrjson.ListTransactionsResult) {
  Account: (string) "",
  Address: (string) (len=35) "SsdLk37kzNfTtvk4nmxFZawcYvuehiCxAPe",
  Amount: (float64) -2.4,
  BlockHash: (string) "",
  BlockIndex: (*int64)(<nil>),
  BlockTime: (int64) 0,
  Category: (string) (len=4) "send",
  Confirmations: (int64) 0,
  Fee: (*float64)(0xc0421aed18)(-0.000253),
  Generated: (bool) false,
  InvolvesWatchOnly: (bool) false,
  Time: (int64) 1535622117,
  TimeReceived: (int64) 1535622117,
  TxID: (string) (len=64) "93676ace2552d5947abdf7ba8e6ca43a4719349dbd23dafcc24b966d5fd6e31d",
  TxType: (*dcrjson.ListTransactionsTxType)(0xc04225c370)((len=7) "regular"),
  Vout: (uint32) 1,
  WalletConflicts: ([]string) {
  },
  Comment: (string) "",
  OtherAccount: (string) ""
 }
```

**Comment by JFixby** [comment]
`TestListTransactions` is fixed, no need to work on it.

File: rpctest/02_validateaddress_test.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+			strings.Compare("imported", acct) != 0 {
+			err := r.WalletRPCClient().CreateNewAccount(acct)
+			if err != nil {
+				t.Fatalf("Unable to create account %s: %v", acct, err)
+			}
+		}
+
+		// Get a new address from current account
+		addr, err := wcl.GetNewAddressGapPolicy(
+			acct, rpcclient.GapPolicyIgnore)
```
Comment (JFixby) [inline rpctest/02_validateaddress_test.go]: Seems like address requested with `GapPolicyIgnore` fails the `IsMine` check at line 61. Is it a bug or expected behavior?

**Comment by JFixby** [comment]
Only 3 tests left failing:

1. `TestValidateAddress` found a bug in `dcrwallet` when `wallet RPC` incorrectly identifies wallet address as `NotMine`
```
02_validateaddress_test.go:123: Address[0] incorrectly identified as NOT mine:
SsrPNpSPoVNV7ontVaHkLNkdDfGxiaw5n7q for <default>
```

2. `TestGetStakeInfo` fails saying:
```
15_getstakeinfo_test.go:77: Stake difficulty mismatch:
0.000200 vs 0.000000 (getstakedifficulty, getstakeinfo)
```

3. `TestPurchaseTickets` runs in endless loop.
`dcrwallet` log also reports:
```
Ticket purchasing failed: wallet.NextStakeDifficultyAfterHeader:
inactive deployment: DCP0001 is not known to be active
```

I will leave it as is for now. 

Tests status:

image

**Comment by jrick** [comment]
Why do all of the test source files now begin with 01, 02...?

**Comment by JFixby** [comment]
The enumeration preserves execution order of the original tests. Original tests were written requiring it.

**Comment by jrick** [comment]
in that case, please revert the changes you made which exported the test functions, and move them all back into a single file in the order they are run.  All exported tests should pass regardless of the order, and it's easier to read and follow the tests as sequential tests when they are all in the same file.  It's not a great test setup, but this is less confusing then relying on undefined behavior of `go test`.

**Comment by JFixby** [comment]
These are not sequential tests. They were written to run sequentially, but that was side effect of using `for`-loop.
Current tests suppose to pass regardless of the order.
I can remove the enumeration but I will not merge them back, because it is an opposite of added value here. The purpose of this PR is to improve code, not to degrade it.

**Comment by jrick** [comment]
If the updated tests do not need to be run in any particular order, then the enumeration should be removed.

Finally, there's no reason to separate the tests like that.  This isn't java, and it's easier to deal with a single file, especially when each of the tests are as short as they are:

```
$ wc -l *
    164 01_getnewaddress_test.go
    160 02_validateaddress_test.go
    139 03_walletpassphrase_test.go
    144 04_getbalance_test.go
    165 05_listaccounts_test.go
    168 06_listunspent_test.go
     92 07_sendtoaddress_test.go
    178 08_sendfrom_test.go
    172 09_sendmany_test.go
    354 10_listtransactions_test.go
    115 11_getsetrelayfee_test.go
    117 12_getsetticketfee_test.go
    121 13_gettickets_test.go
    206 14_purchasetickets_test.go
    246 15_getstakeinfo_test.go
     30 16_walletinfo_test.go
```

Not splitting also reduces the size of the diff and makes reviewing this easier.  Do not make unnecessary, trivial changes such as this; I won't allow it.

**Comment by JFixby** [comment]
I didn't really change the testcaces in the first commit, just moved them into dedicated files.
So you can see the actual diffs in the following commits here.
I can temporary merge tests so you can see the diffs, but then we should split them back.

Also, it is precisely my goal here: to keep all files below 200 lines of code. 
Not because of Java, but for the purpose of code readability and maintainability, meaning:
Simpler future diffs.
Easier to find required function and a struct with all it's methods (by file name).
You can simply copy file you like to another project/package.
One test - one file.
Files are still readable one or two years later. Even when you completely forgot them.
Easier to spot a bug. Spotting a bug in longer files is hard, and is impossible for files above 400 lines.

I care about my fellow contributors and aim to make make their life easier. That is why I generally keep files below 200LOC.

I can of course merge the whole thing back into a few giant files. But I don't know why you want me to do this to you. Just tell me you are sure about this and I'll merge it.

**Comment by jrick** [comment]
Yes, please do (for all of the tests files currently prefixed with numbers).  You did not discuss this change in our dev chat and tried to include it in an unrelated change (the goal of which was make the tests work).  This does not follow our development policy.

Long files are not a problem, but too many smaller files is a pain to work with.  Splitting code into multitudes of small files is not a worthwhile use of your dcrwallet development time, so please abstain from this in the future.

File: rpctest/rpcharness_test.go [hunk 3]
```diff
@@ -937,7 +1112,7 @@ func testSendToAddress(r *Harness, t *testing.T) {
 	// Check all inputs
 	for i, txIn := range rawTx.MsgTx().TxIn {
 		prevOut := &txIn.PreviousOutPoint
-		t.Logf("Checking previous outpoint %v, %v", i, prevOut.String())
+		//t.Logf("Checking previous outpoint %v, %v", i, prevOut.String())
```
Comment (vctt94) [inline rpctest/rpcharness_test.go]: can remove this log

File: rpctest/rpcharness_test.go [hunk 4]
```diff
@@ -1159,30 +1374,50 @@ func testSendMany(r *Harness, t *testing.T) {
...
 
 	// Get rawTx of sent txid so we can calculate the fee that was used
-	rawTx, err := r.Node.GetRawTransaction(txid)
+	rawTx, err := r.DcrdRPCClient().GetRawTransaction(txid)
 	if err != nil {
 		t.Fatalf("getrawtransaction failed: %v", err)
 	}
 	fee := getWireMsgTxFee(rawTx)
-	t.Log("Raw TX before mining block: ", rawTx, " Fee: ", fee)
+	//t.Log("Raw TX before mining block: ", rawTx, " Fee: ", fee)
```
Comment (vctt94) [inline rpctest/rpcharness_test.go]: can remove

File: rpctest/rpcharness_test.go [hunk 5]
```diff
@@ -1557,7 +1848,7 @@ func testGetSetRelayFee(r *Harness, t *testing.T) {
 	feeRate := fee.ToCoin() / float64(rawTx.MsgTx().SerializeSize()) * 1000
 
 	// Ensure actual fee is at least nominal
-	t.Logf("Set relay fee: %v, actual: %v", walletInfo.TxFee, feeRate)
+	//t.Logf("Set relay fee: %v, actual: %v", walletInfo.TxFee, feeRate)
```
Comment (vctt94) [inline rpctest/rpcharness_test.go]: can remove

File: rpctest/rpcharness_test.go [hunk 6]
```diff
@@ -1656,7 +1955,7 @@ func testGetSetTicketFee(r *Harness, t *testing.T) {
 	feeRate := fee.ToCoin() / float64(rawTx.MsgTx().SerializeSize()) * 1000
 
 	// Ensure actual fee is at least nominal
-	t.Logf("Set ticket fee: %v, actual: %v", nominalTicketFee, feeRate)
+	//t.Logf("Set ticket fee: %v, actual: %v", nominalTicketFee, feeRate)
```
Comment (vctt94) [inline rpctest/rpcharness_test.go]: can remove

File: rpctest/rpcharness_test.go [hunk 7]
```diff
@@ -340,13 +389,96 @@ func testValidateAddress(r *Harness, t *testing.T) {
 	if validRes.IsMine {
 		t.Fatalf("Address incorrectly identified as mine: %s", devSubAddrStr)
 	}
-	// for ismine==false, nothing else to test
+	//-----------------------------------------
```
Comment (vctt94) [inline rpctest/rpcharness_test.go]: no need to this comment

**Comment by JFixby** [comment]
@vctt94 you are welcome to rebase your branch.

**Comment by njirap** [comment]
screen shot 2018-12-11 at 11 51 46

Hi @JFixby on checking out your PR these tests still fail. How can I replicate the fix from my end?

Also this PR fixes the `Unable to setup test chain:  connection timed out: x509: certificate signed by unknown authority (possibly because of "x509: ECDSA verification failure" while trying to verify candidate authority certificate` issue, any chance it could be merged soon? Thanks

**Comment by njirap** [comment]
Thanks alot @JFixby 

I am guided

**Comment by jrick** [comment]
rpctest has been removed in a865b1fc923d0d21b7c57974adf83e63548c3693.

**Pull request closed without merging**